Merge settings one place instead of five places

ryanofsky opened the pull request:
This is a refactoring-only change that makes it easier to add a new settings source.

This PR doesn't change behavior. The [`util_ArgsMerge`](https://github.com/bitcoin/bitcoin/blob/deb2327b435925c6a39ca654a79283b8eb6aeb86/src/test/util_tests.cpp#L626-L822) and [`util_ChainMerge`](https://github.com/bitcoin/bitcoin/blob/deb2327b435925c6a39ca654a79283b8eb6aeb86/src/test/util_tests.cpp#L843-L924) tests added in #15869 and #15988 were written specifically to confirm that ArgsManager settings are parsed, merged, and returned the same way before and after this change.

This change:

- Makes it easier to add new settings sources that can get merged with existing sources (see 70675c3e4975203ad6222ba2b00c83b4e4213793 from #15935).
- Separates parsing of settings from merging of settings, and deduplicates merging code so it doesn't happen five different places ([GetArg](https://github.com/bitcoin/bitcoin/blob/c459c5f70176928adcee4935813a2dbe7f4dbd51/src/util/system.cpp#L221-L244), [GetNetBoolArg](https://github.com/bitcoin/bitcoin/blob/c459c5f70176928adcee4935813a2dbe7f4dbd51/src/util/system.cpp#L255-L261), [GetArgs](https://github.com/bitcoin/bitcoin/blob/c459c5f70176928adcee4935813a2dbe7f4dbd51/src/util/system.cpp#L460-L467), [IsArgNegated](https://github.com/bitcoin/bitcoin/blob/c459c5f70176928adcee4935813a2dbe7f4dbd51/src/util/system.cpp#L482-L491), [GetUnsuitableSectionOnlyArgs](https://github.com/bitcoin/bitcoin/blob/c459c5f70176928adcee4935813a2dbe7f4dbd51/src/util/system.cpp#L343-L352)) in inconsistent ways.
- Documents and tests current strange merging behaviors, so they be cleaned up in the future if resulting code simplifications and UX improvements warrant loss of backwards compatibility. The newly documented behaviors are: command line [ignored arguments](https://github.com/ryanofsky/bitcoin/blob/69d44f3cc75a68d404ca0e1ca2b4831fd2bac4bb/src/util/system.cpp#L323-L326) and [more ignored arguments](https://github.com/ryanofsky/bitcoin/blob/69d44f3cc75a68d404ca0e1ca2b4831fd2bac4bb/src/util/settings.cpp#L67-L72), and config file [reverse precedence](https://github.com/ryanofsky/bitcoin/blob/69d44f3cc75a68d404ca0e1ca2b4831fd2bac4bb/src/util/settings.cpp#L61-L65), [inconsistently applied top-level settings](https://github.com/ryanofsky/bitcoin/blob/69d44f3cc75a68d404ca0e1ca2b4831fd2bac4bb/src/util/settings.cpp#L55-L59), and [zombie values](https://github.com/ryanofsky/bitcoin/blob/69d44f3cc75a68d404ca0e1ca2b4831fd2bac4bb/src/util/settings.cpp#L101-L108).

The original motivation for this change was to make it easy to add a new persistent setting source without introducing more bugs and inconsistencies. Two commits building on top of this to add a persistent `-wallet` setting are pretty straightforward and show how the new code can be extended:

* 70675c3e4975203ad6222ba2b00c83b4e4213793 from #15935 – _Add \<datadir>/settings.json persistent settings storage_
* 04c80c40df9fc6f4734ba238ea7f65607cf88089 from #15937 – _Add loadwallet and createwallet RPC load_on_startup options_

ryanofsky committed:
Add util::Settings struct and helper functions.

Implement merging of settings from different sources (command line and config
file) separately from parsing code in system.cpp, so it is easier to add new
sources.

Document current inconsistent merging behavior without changing it.

This commit only adds new settings code without using it. The next commit calls
the new code to replace existing code in system.cpp.

Co-authored-by: John Newbery <john@johnnewbery.com>

ryanofsky committed:
Deduplicate settings merge code

Get rid of settings merging code in util/system.cpp repeated 5 places,
inconsistently:

- ArgsManagerHelper::GetArg
- ArgsManagerHelper::GetNetBoolArg
- ArgsManager::GetArgs
- ArgsManager::IsArgNegated
- ArgsManager::GetUnsuitableSectionOnlyArgs

Having settings merging code separated from parsing simplifies parsing somewhat
(for example negated values can simply be represented as false values instead
of partially cleared or emply placeholder lists).

Having settings merge happen one place instead of 5 makes it easier to add new
settings sources and harder to introduce new inconsistencies in the way
settings are merged.

This commit does not change behavior in any way.

promag commented on src/util/system.cpp:
```diff
@@ -163,103 +164,34 @@ static bool InterpretBool(const std::string& strValue)
     return (atoi(strValue) != 0);
 }
 
+static bool ValidArgStr(const std::string& arg)
```
nit, `IsValidArgStr`?

promag commented on src/util/system.cpp:
```diff
@@ -163,103 +164,34 @@ static bool InterpretBool(const std::string& strValue)
 
+static bool ValidArgStr(const std::string& arg)
+{
+    return !arg.empty() && arg[0] == '-';
```
Maybe `arg.size() > 1 && arg[0] == '-'` or do we need `-`?

promag reviewed (commented):
I had a refactor (which I did't submit) that supported chaining `ArgsManager`. The idea was to support changing some args when calling some RPC, so a `ArgsManager` is created with the "overridden" args and passed thru. Is this something you are considering supporting or do you see a different approach?

Concept ACK.

ryanofsky committed:
Add settings_tests

Co-authored-by: James O'Beirne <james.obeirne@gmail.com>

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/15934#pullrequestreview-232997353 from promag

> I had a refactor (which I did't submit) that supported chaining ArgsManager. The idea was to support changing some args when calling some RPC, so a ArgsManager is created with the "overridden" args and passed thru. Is this something you are considering supporting or do you see a different approach?

This change does make it easier to add new settings sources (with consistent handling of negated args and things), so it should be compatible with your idea and maybe helpful.

Depending on the situation, I think having chained or scoped settings could be a good idea or not. I do think that in wallet code and application code generally it's good to get away from using key-value storage classes like `ArgsManager` or `UniValue` as quickly as possible, and switch to more direct representations like `CCoinControl` that are type safe and can be accessed more simply.

fanquake commented:
Concept ACK

jamesob reviewed (approved):
Tested ACK https://github.com/bitcoin/bitcoin/pull/15934/commits/1d543adae593bdbfd954e80ed61ac907db0c5a7b

Generated and reviewed the test output locally. Mucked around with various argument formulations using the following config file:
```
dbcache=100
[main]
dbcache=200
[test]
dbcache=300
```
and commandline invocations e.g.
```sh
./src/bitcoind -conf=$(pwd)/test.conf -dbcache=1000 -dbcache=500 | grep Using
```
to verify dbcache being set as expected.

---

This is a well-written change that cleans up a lot of gnarly, duplicated settings munging. It explicitly outlines surprising corner cases in existing behavior (with docs too), and makes reasoning about settings merge order easier. This change also introduces substantial test coverage to settings management (`util::Settings`).

After this is merged, adding a read-write settings file (whether it's JSON or something else) will be much easier.

MarcoFalke commented on src/test/util.h:
```diff
@@ -34,5 +34,37 @@ std::string getnewaddress(CWallet& w);
+        }
+        if (!skip_string) fn();
+    }
+}
```
So it should be rebased, imo. Otherwise `git blame` will return different results for the same lines depending on what commit is currently checked out.

ryanofsky reviewed (commented):
Updated 1d543adae593bdbfd954e80ed61ac907db0c5a7b -> 2dfeff1e6841b77e4e689d54c08557b938407ba1 ([pr/mergeset.6](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.6) -> [pr/mergeset.7](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.7)), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mergeset.6..pr/mergeset.7)) with suggested changes.
Rebased 2dfeff1e6841b77e4e689d54c08557b938407ba1 -> 955c782eb776669576a798122da6557fcca8ee39 ([pr/mergeset.7](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.7) -> [pr/mergeset.8](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.8)) to share common code with #15988.

jamesob commented:
re-tACK https://github.com/bitcoin/bitcoin/pull/15934/commits/955c782eb776669576a798122da6557fcca8ee39 based on [the interdiff]( https://gist.github.com/jamesob/d8a458334386ab67082956e404fb704d) and running an abbreviated version of the testing above.

ryanofsky commented:
Rebased 955c782eb776669576a798122da6557fcca8ee39 -> 14a6dfcb929d2313577788765f7dc47dd98afbe9 ([pr/mergeset.8](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.8) -> [pr/mergeset.9](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.9)) due to conflict with #16278

jamesob commented:
reACK https://github.com/bitcoin/bitcoin/pull/15934/commits/14a6dfcb929d2313577788765f7dc47dd98afbe9 based on [interdiff](https://gist.github.com/jamesob/0b4adf37c627e6171c2599dfa95f0901). Only change since `pr/mergeset.8` is a [trivial `LogPrintf` fix](https://gist.github.com/jamesob/0b4adf37c627e6171c2599dfa95f0901#file-8-to-9-diff-L72-L73).

ryanofsky commented:
Rebased 14a6dfcb929d2313577788765f7dc47dd98afbe9 -> d074e431b76f156e94eb7ed2601d5af4fb65c6cb ([pr/mergeset.9](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.9) -> [pr/mergeset.10](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.10)) due to conflict with #16300

jamesob commented:
reACK d074e431b76f156e94eb7ed2601d5af4fb65c6cb based on [an empty interdiff](https://gist.github.com/jamesob/fea5597b189e899154e1e9a9bdca0d94).

hebasto commented:
Concept ACK.

ariard commented on src/util/settings.cpp:
```diff
@@ -0,0 +1,157 @@
+static void MergeSettings(const Settings& settings, const std::string& section, const std::string& name, Fn&& fn)
+{
+    if (auto* value = FindKey(settings.forced_settings, name)) {
+        fn(Source(SettingsSpan(value)).SetForced());
```
nit: any reason why you don't dereference compare to others?

ryanofsky replied:
> nit: any reason why you don't dereference compare to others?

sss

ariard reviewed (commented):
Just a few nits on the new code to clarify some points.

Next step is to take every weird compatibility spotted and check if it still works the same on both master nodes and patchset nodes. And look in depth to existing code if there is no uncovered one.

Maybe they should be documented in one place, like at the beginning of `ArgsManagerHelper` (or `src/doc/bitcoin-conf.md` ?) with general rules of precedence/syntax and all the exceptions. 

Have been through the #15935 debate, IMO I really like ability to document my config values to remember how to  fine-tune them at later time. Not sure if the distinction between config file = sysadmin, dynamic settings = GUI users will hold, e.g I would like to call `pruneblockchain` and get my config updated for next time I reboot the system not having my disks blow up. Beyond that, don't have hard opinion on the matter.

ryanofsky commented on src/util/settings.h:
```diff
@@ -0,0 +1,72 @@
+
+namespace util {
+
+//! Settings value type (string/integer/boolean/null variant).
```
> Also key-value or array as requirements ? I suppose any other variant type should be easy to use with basic types..

ddd

ryanofsky replied:
These definitely aren't not requirements of the type right now. There is no syntax for these things. Maybe this would be useful in the future for something like per-wallet settings (different preferences, RPC authentication setting), and I guess this is a little easier now, but otherwise the PR is neutral on this.

ryanofsky replied:
> I suppose any other variant type should be easy to use with basic types..

I think `boost::variant` would be an inferior choice, but just as an experiment I tried dropping it in here with some helper methods to make it work without changes to other code:

```c++
struct SettingsValue : public boost::variant<std::nullptr_t, bool, int64_t, std::string>
{
    using variant::variant;
    template<typename T> const T* ptr() const { return boost::get<T>(this); }
    bool isNull() const { return ptr<std::nullptr_t>(); }
    bool isBool() const { return ptr<bool>(); }
    bool isNum() const { return ptr<int64_t>(); }
    bool isFalse() const { return ptr<bool>() && !*ptr<bool>(); }
    bool isTrue() const { return ptr<bool>() && *ptr<bool>(); }
    const std::string& get_str() const { return *ptr<std::string>(); }
    const bool& get_bool() const { return *ptr<bool>(); }
    const int64_t& get_int64() const { return *ptr<int64_t>(); }
};
```

This was sufficient for everything except `settings_test.cpp` (which uses serialization), so it confirms `SettingsValue` only requires storing ints, bools, strings and nulls, and the choice of type isn't very important.

jnewbery replied:
> I just listed the required methods and added a github link, but I can add the example code inline if preferred.

Looks great to me as it is. Thanks!

ryanofsky commented on src/util/settings.h:
```diff
@@ -0,0 +1,72 @@
+//! Get settings value from combined sources: forced settings, command line
+//! arguments and the read-only config file.
+//!
+//! @param ignore_default_section_config - ignore values set in the top-level
```
> what's top-level section of the config file ? The non-network one?

ddd

ryanofsky commented on src/util/settings.h:
```diff
@@ -0,0 +1,72 @@
+
+//! Get combined setting value similar to GetSetting(), except if setting was
+//! specified multiple times, return a list of all the values specified.
+std::vector<SettingsValue> GetListSetting(const Settings& settings, const std::string& section, const std::string& name, bool ignore_default_section_config);
```
> nit: `GetSettingAllOccurence`?, given name was expecting to query multiple settings at same time

ddd

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/15934#discussion_r335798710

> > nit: `GetSettingAllOccurence`?, given name was expecting to query multiple settings at same time
> 
> Sure, will rename.

Forgot to do this last time. I think I'll use `GetSetting` for single valued settings and `GetSettingsList` for many valued settings. I'd like to keep using the word "list" here to be consistent with #16545 which also uses it with `TYPE_STRING_LIST` as one of the settings types.

ryanofsky reviewed (commented):
> Next step is to take every weird compatibility spotted and check if it still works the same on both master nodes and patchset nodes. And look in depth to existing code if there is no uncovered one.

I'm not sure you're suggesting. Maybe you could go into more detail. I added an extensive unit test in #15869 specifically to support this PR and check all these corner cases and make sure there is no loss of compatibility. I would like to get rid of the confusing behaviors by turning ambiguous configurations into errors, and I implemented flags to start to do this in #16545.

> Maybe they should be documented in one place, like at the beginning of `ArgsManagerHelper` (or `src/doc/bitcoin-conf.md` ?) with general rules of precedence/syntax and all the exceptions.

I want to make ambiguous configurations into errors, and fix cases where settings are parsed in nonsensical ways. I tried to write clear code comments, and am happy to improve them, but I don't want to make nonsensical behaviors permanent or document them for end users.

> Have been through the #15935 debate, IMO I really like ability to document my config values to remember how to fine-tune them at later time. Not sure if the distinction between config file = sysadmin, dynamic settings = GUI users will hold, e.g I would like to call `pruneblockchain` and get my config updated for next time I reboot the system not having my disks blow up. Beyond that, don't have hard opinion on the matter.

Thanks, and it'd be helpful if you described your use-case with as many specifics as possible to #15935. But there is definitely nothing in this PR that dictates use of any storage format or commenting style, so I think that discussion is not on topic here.

ryanofsky commented:
Updated 61e034f1f6b3beeeed51ef8aaf7762f9244a7c8f -> 40593ecc0bc48b22054da73eb4bdf1b3aeaedd4f ([pr/mergeset.15](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.15) -> [pr/mergeset.16](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.16), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mergeset.15..pr/mergeset.16)) just updating some comments discussed above and passing by reference instead of pointer to a span constructor

ariard commented on src/util/system.cpp:
```diff
@@ -451,69 +358,42 @@ unsigned int ArgsManager::FlagsOfKnownArg(const std::string& key) const
 std::vector<std::string> ArgsManager::GetArgs(const std::string& strArg) const
 {
-    std::vector<std::string> result = {};
-    if (IsArgNegated(strArg)) return result; // special case
```
Just to be sure the `IsArgNegated` is preserved here in `GetListSetting` via `source.span.negated() > 0` ?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/15934#discussion_r337809469

> Just to be sure the `IsArgNegated` is preserved here in `GetListSetting` via `source.span.negated() > 0` ?

It's handled implicitly by `for (const auto& value : source.span)` in `GetListSetting`, because the `SettingsSpan` iterator skips negated values in its `begin()` method.

The `source.span.negated() > 0` check you're referring is used after that to set the `result_complete` variable, which is used to apply settings precedence
`command_line_options` > `rw_settings` > `ro_settings`, and make sure for example that a negation on the command line will clear out any lower priority settings in `ro_config` and `rw_settings`. I'll add a comment to say this explicitly on the result_complete line.

ariard commented on src/util/system.cpp:
```diff
@@ -399,49 +312,43 @@ bool ArgsManager::ParseParameters(int argc, const char* const argv[], std::strin
+            // line options with section prefixes are allowed but ignored. It
+            // would be better if these options triggered the Invalid parameter
+            // error below.
+            if (section.empty()) {
```
Hmmm I can't get from where this behavior is extracted ?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/15934#discussion_r337810134

> Hmmm I can't get from where this behavior is extracted ?

This behavior makes no sense and was never intended, so it's buried away in current code. It happens because of the following code in `ArgsManager::ParseParameters`:

https://github.com/bitcoin/bitcoin/blob/9dd6bbba613d7462afdb6276c4002bc183478528/src/util/system.cpp#L411-L420

A section `section.name` command line setting here will fail to trigger the "Invalid parameter" error because `FlagsOfKnownArg` strips out the `section.` prefix and finds that `name` is a known setting. But `InterpretOption` fails to strip out the `section.` prefix and stores the setting in the `m_override_args` map under `section.name` instead of `name`, where it is ignored and never has any effect.

The new code improves this by documenting the behavior and making it easy to change in the future. It also avoids bugs like this by splitting the `section.name` string exactly once in `InterpretOption` instead of many different places and times.

ariard reviewed (approved):
Code review and tested ACK 40593ec

ryanofsky reviewed (commented):
Thanks for the review!

fjahr commented on src/util/system.cpp:
```diff
@@ -451,69 +358,42 @@ unsigned int ArgsManager::FlagsOfKnownArg(const std::string& key) const
+    std::vector<std::string> result;
+    for (const util::SettingsValue& value :
+        util::GetListSetting(m_settings, m_network, SettingName(strArg), ignore_default_section_config)) {
+        result.push_back(value.isFalse() ? "0" : value.isTrue() ? "1" : value.get_str());
```
On second thought: since these nested ternary one-liners are repeated multiple times, I would suggest a helper method that converts `value` to whatever gets pushed into `result`

fjahr reviewed (approved):
tested ACK 40593ecc0bc48b22054da73eb4bdf1b3aeaedd4f

Reviewed code, ran tests, did some manual testing, cherry-picked the regression test to verify results are the same on current master.

Thank you for this nice cleanup, I really like the regression test. Just had some nit about the readability with the ternary statements in `system.cpp`.

Edit: I also found the changes in `system.cpp` hard to review in one. Maybe they could have been split up into different commits.

jonatack commented:
Tested ACK 40593ecc0bc48b22054da73eb4bdf1b3aeaedd4f. Will review and re-test 8ca6338e994000c82c813721c85ebafd61250641 tomorrow.

jonatack reviewed (commented):
Tested ACK 40593ecc0bc48b22054da73eb4bdf1b3aeaedd4f. Will review and test the latest version at 8ca6338e994000c82c813721c85ebafd61250641 tomorrow.

Talkless commented on src/test/util_tests.cpp:
```diff
@@ -156,19 +157,58 @@ BOOST_AUTO_TEST_CASE(util_FormatISO8601Date)
     std::map<std::string, std::vector<std::string> >& GetOverrideArgs() { return m_override_args; }
     std::map<std::string, std::vector<std::string> >& GetConfigArgs() { return m_config_args; }
+    bool ParseParameters(int argc, const char* const argv[], std::string& error) {
+        bool result = ArgsManager::ParseParameters(argc, argv, error);
```
**const** bool would help here, it actually too a wile to see that it is not changed throughout this function.

Talkless commented on src/test/util_tests.cpp:
```diff
@@ -156,19 +157,58 @@ BOOST_AUTO_TEST_CASE(util_FormatISO8601Date)
+        LOCK(cs_args);
+        for (const auto* args : {&m_override_args, &m_config_args}) {
+            for (const auto& arg : *args) {
+                std::string name = arg.first[0] == '-' ? arg.first.substr(1) : arg.first;
```
could be `const std::string`

Talkless commented on src/util/settings.cpp:
```diff
@@ -0,0 +1,161 @@
+    bool config_file_top_level = false;
+
+    Source(SettingsSpan span) : span(span) {}
+    Source& SetForced() { forced = true; return *this; }
```
`noexcept`?

Talkless commented on src/util/settings.cpp:
```diff
@@ -0,0 +1,161 @@
+
+    Source(SettingsSpan span) : span(span) {}
+    Source& SetForced() { forced = true; return *this; }
+    Source& SetConfigFile(bool top_level) { config_file = true; config_file_top_level = top_level; return *this; }
```
could be `noexcept`

ryanofsky commented:
Updated 40593ecc0bc48b22054da73eb4bdf1b3aeaedd4f -> 8ca6338e994000c82c813721c85ebafd61250641 ([pr/mergeset.16](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.16) -> [pr/mergeset.17](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.17), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mergeset.16..pr/mergeset.17)) with minor cleanups

Sorry for the churn. Only changes were the comments and cleanups I promised Antoine this morning

Talkless commented on src/util/settings.cpp:
```diff
@@ -0,0 +1,161 @@
+            }
+        }
+    }
+    SettingsSpan span;
```
Consider marking as `const` and applying [ES.28: Use lambdas for complex initialization, especially of const variables](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es28-use-lambdas-for-complex-initialization-especially-of-const-variables). That re-assignment `span = SettingsSpan(*values);` is rather "hidden".

Talkless commented on src/util/settings.cpp:
```diff
@@ -0,0 +1,161 @@
+        if (skip_negated_command_line && source.span.last_negated()) return;
+
+        // Stick with highest priority value, keeping result if already set.
+        if (!result.isNull()) return;
```
This looks redundant. `result` is always null here, I don't see where it can be changed above. All this could be rewritten as a `return <something>` instead of changing via lambda capturer reference...

ryanofsky replied:
I can add a 

> This looks redundant. `result` is always null here, I don't see where it can be changed above. All this could be rewritten as a `return <something>` instead of changing via lambda capturer reference...

I can maybe change the comment above to clarify, but `return` here is what keeps lower priority values from overwriting higher priority values,

Talkless replied:
Yep, I've missed that fact that lambda is called multiple times, thanks for clarification. So this comment is moot.

Talkless commented on src/util/settings.cpp:
```diff
@@ -0,0 +1,161 @@
+
+SettingsSpan::SettingsSpan(const std::vector<SettingsValue>& vec) : SettingsSpan(vec.data(), vec.size()) {}
+const SettingsValue* SettingsSpan::begin() const { return data + negated(); }
+const SettingsValue* SettingsSpan::end() const { return data + size; }
```
`end()` could be `noexcept`, not sure about all these other trivial functions (with regards to `isFalse()` via `negated()`).

Talkless commented on src/util/settings.h:
```diff
@@ -0,0 +1,82 @@
+struct SettingsSpan {
+    SettingsSpan(const SettingsValue* data, size_t size) : data(data), size(size) {}
+    SettingsSpan(const std::vector<SettingsValue>& vec);
+    SettingsSpan(const SettingsValue& value) : SettingsSpan(&value, 1) {}
```
explcit? noexcept?

Talkless commented on src/util/settings.h:
```diff
@@ -0,0 +1,82 @@
+//! Iterable list of settings that skips negated values.
+struct SettingsSpan {
+    SettingsSpan(const SettingsValue* data, size_t size) : data(data), size(size) {}
+    SettingsSpan(const std::vector<SettingsValue>& vec);
```
Missing `exclicit`? Do we really need to support `SettigsSpan s = std::vec...` ?

Talkless replied:
Oh no, I do not suggest to remove this helper constructor, maybe just adding `explicit` for that constructor. Currently it implicilty-converts from std::vecotor:
```
//...
DoSomething(const SettingsSpan& span);
//...
std::vector<SettingsValue>& vec;
DoSomething(vec); // implicilty conerts to SettingsSpan, kinda "sneaky".
```
With `explicit` constructor:
```
DoSomething(SettingsValue{vec}); // we see that we pass SettngsValue, not "just" a vector
```

Although looking at `std::span` implementations, these constructors are also implicit there, so I guess you can ignore this comment.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/15934#discussion_r338191129

> Although looking at `std::span` implementations, these constructors are also implicit there, so I guess you can ignore this comment.

Yes, it's a perfectly safe conversion that doesn't create ownership or performance issues, and in case of `std::span` reduces noise and boilerplate in code. But in any case I updated this PR to use explicit here the other day based on your feedback (https://github.com/ryanofsky/bitcoin/compare/pr/mergeset.17..pr/mergeset.18) and I think it's an improvement.

Talkless commented on src/util/settings.h:
```diff
@@ -0,0 +1,82 @@
+    SettingsSpan(const SettingsValue* data, size_t size) : data(data), size(size) {}
+    SettingsSpan(const std::vector<SettingsValue>& vec);
+    SettingsSpan(const SettingsValue& value) : SettingsSpan(&value, 1) {}
+    SettingsSpan() {}
```
maybe just `= default` ?

Talkless commented on src/util/settings.h:
```diff
@@ -0,0 +1,82 @@
+template <typename Map, typename Key>
+auto FindKey(Map&& map, Key&& key) -> decltype(&map.at(key))
+{
+    auto it = map.find(key);
```
`const auto it` as `it` is not changed later. 

P.S. really sorry for not using "Review Changes" feature. I've been informed that these "add single comment" spams the notification channel... Will do better next time.

ryanofsky reviewed (commented):
Thanks for the reviews!

Updated 8ca6338e994000c82c813721c85ebafd61250641 -> 86aff9bda31c9d60cc477700da4600ad5930bcda ([pr/mergeset.17](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.17) -> [pr/mergeset.18](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.18), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mergeset.17..pr/mergeset.18)) just adding suggested const/noexcept/explicits.

re: https://github.com/bitcoin/bitcoin/pull/15934#pullrequestreview-305976230

> Just had some nit about the readability with the ternary statements in system.cpp.

I'd be fine with changing these statements, but this is just a convention of abbreviating:

```c++
if (condition1)
   value = result1;
else if (condition2)
   value = result2;
else
   value = fallback;
```

to:

```c++
value = condition1 ? result1 : condition2 ? result2 : fallback;
```

which I think is common in many languages (though [famously broken in PHP](https://bugs.php.net/bug.php?id=61915)).

> I also found the changes in system.cpp hard to review in one. Maybe they could have been split up into different commits.

I'm not sure if there's something that can be done about this. It's the switch from old settings storage to new storage that requires all the changes:

```diff
-    std::map<std::string, std::vector<std::string>> m_override_args GUARDED_BY(cs_args);
-    std::map<std::string, std::vector<std::string>> m_config_args GUARDED_BY(cs_args);
+    util::Settings m_settings GUARDED_BY(cs_args);
```

fjahr commented:
ACK 86aff9bda31c9d60cc477700da4600ad5930bcda

Confirmed that diff is only fixing nits as discussed in reviews. Re-ran tests. Some manual testing.

Talkless commented on src/util/settings.h:
```diff
@@ -0,0 +1,82 @@
+    explicit SettingsSpan() = default;
+    explicit SettingsSpan(const SettingsValue& value) noexcept : SettingsSpan(&value, 1) {}
+    explicit SettingsSpan(const SettingsValue* data, size_t size) noexcept : data(data), size(size) {}
+    explicit SettingsSpan(const std::vector<SettingsValue>& vec) noexcept;
```
Now there are some redundant `explicit`'s. There's no need to mark `SettingsSpan() = default;` or more-than-one-argument-taking `SettingsSpan(const SettingsValue* data, size_t size)` as `explicit`. It's only for the cases when constructor can initialize with _single_ argument.

Talkless commented on src/util/settings.cpp:
```diff
@@ -0,0 +1,161 @@
+        }
+    }
+    SettingsSpan span;
+    if (auto* map = FindKey(settings.ro_config, "")) {
```
I believe `""` are rather overused on codebases. This invokes `std::string`'s copy constructor taking `const char*` that will try to `strlen()`, etc. Creating default-constructed `std::string{}`/ `std::string()`/`{} (in right context)` would feel more "pedantically correct" :) .

Talkless replied:
Right, I'm used to QString too much. std::string optimizes well ("" same as {}), it's QString where it does add cost (extra function all). Sorry for bothering.

Talkless commented on src/util/system.cpp:
```diff
@@ -399,49 +312,43 @@ bool ArgsManager::ParseParameters(int argc, const char* const argv[], std::strin
+    bool success = true;
+    if (auto* includes = util::FindKey(m_settings.command_line_options, "includeconf")) {
+        for (const auto& include : util::SettingsSpan(*includes)) {
+            error += "-includeconf cannot be used from commandline; -includeconf=" + include.get_str() + "\n";
```
`"\n"` -> `'\n'`, no need for const char* / std::string for single character.

Talkless commented on src/util/system.cpp:
```diff
@@ -399,49 +312,43 @@ bool ArgsManager::ParseParameters(int argc, const char* const argv[], std::strin
     LOCK(cs_args);
     for (const auto& arg_map : m_available_args) {
-        const auto search = arg_map.second.find(base_arg_name);
+        const auto search = arg_map.second.find("-" + key);
```
`"-"` -> `'-'`

Talkless commented on src/util/system.cpp:
```diff
@@ -451,69 +358,42 @@ unsigned int ArgsManager::FlagsOfKnownArg(const std::string& key) const
-    ArgsManagerHelper::AddArgs(result, m_override_args, strArg);
-    if (!m_network.empty()) {
-        ArgsManagerHelper::AddArgs(result, m_config_args, ArgsManagerHelper::NetworkArg(*this, strArg));
+    bool ignore_default_section_config = !ArgsManagerHelper::UseDefaultSection(*this, strArg);
```
const bool?

Talkless commented on src/util/system.cpp:
```diff
@@ -451,69 +358,42 @@ unsigned int ArgsManager::FlagsOfKnownArg(const std::string& key) const
-    std::pair<bool,std::string> found_res = ArgsManagerHelper::GetArg(*this, strArg);
-    if (found_res.first) return found_res.second;
-    return strDefault;
+    util::SettingsValue value = ArgsManagerHelper::Get(*this, strArg);
```
**const** util::SettingsValue value

Talkless commented on src/util/system.cpp:
```diff
@@ -451,69 +358,42 @@ unsigned int ArgsManager::FlagsOfKnownArg(const std::string& key) const
-    std::pair<bool,std::string> found_res = ArgsManagerHelper::GetArg(*this, strArg);
-    if (found_res.first) return atoi64(found_res.second);
-    return nDefault;
+    util::SettingsValue value = ArgsManagerHelper::Get(*this, strArg);
```
**const** util::SettingsValue value

Talkless commented on src/util/system.cpp:
```diff
@@ -451,69 +358,42 @@ unsigned int ArgsManager::FlagsOfKnownArg(const std::string& key) const
-    std::pair<bool,std::string> found_res = ArgsManagerHelper::GetArg(*this, strArg);
-    if (found_res.first) return InterpretBool(found_res.second);
-    return fDefault;
+    util::SettingsValue value = ArgsManagerHelper::Get(*this, strArg);
```
**const** util::SettingsValue value

Talkless commented on src/util/system.cpp:
```diff
@@ -890,25 +773,31 @@ bool ArgsManager::ReadConfigFiles(std::string& error, bool ignore_invalid_keys)
         {
             LOCK(cs_args);
-            emptyIncludeConf = m_override_args.count("-includeconf") == 0;
+            std::vector<util::SettingsValue>* includes = util::FindKey(m_settings.command_line_options, "includeconf");
```
I believe it could be **const** std::vector<util::SettingsValue>* **const** includes. Neither value (vector) nor pointer is ever changed (right?).

Talkless commented on src/util/system.cpp:
```diff
@@ -890,25 +773,31 @@ bool ArgsManager::ReadConfigFiles(std::string& error, bool ignore_invalid_keys)
+            // We haven't set m_network yet (that happens in SelectParams()), so manually check
+            // for network.includeconf args.
+            size_t chain_includes = add_includes(chain_id);
+            size_t default_includes = add_includes("");
```
could be default-constucted instread of copy-constructed std::string: `add_includes({})`;

Also, both `size_t`'s could be `const`.

Talkless commented on src/util/system.cpp:
```diff
@@ -924,15 +813,14 @@ bool ArgsManager::ReadConfigFiles(std::string& error, bool ignore_invalid_keys)
-            includeconf = GetArgs("-includeconf");
+            includeconf.clear();
+            add_includes(chain_id, /* skip= */ chain_includes);
+            add_includes("", /* skip= */ default_includes);
```
`""` -> `{}`  too ?

ryanofsky reviewed (commented):
Updated 86aff9bda31c9d60cc477700da4600ad5930bcda -> fdae2210f458d2c4a74fe0c079501bed39ec6404 ([pr/mergeset.18](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.18) -> [pr/mergeset.19](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.19), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mergeset.18..pr/mergeset.19)) with some suggested minor changes

Talkless commented on src/util/settings.cpp:
```diff
@@ -0,0 +1,161 @@
+const SettingsValue* SettingsSpan::end() const { return data + size; }
+bool SettingsSpan::empty() const { return size == 0 || last_negated(); }
+bool SettingsSpan::last_negated() const { return size > 0 && data[size - 1].isFalse(); }
+size_t SettingsSpan::negated() const
```
`std::count_if` would be simplest, but not sure if semantics will be the same, i.e. will there acutally be N sequental `.isFalse()`'es.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/15934#discussion_r338987908

> `std::count_if` would be simplest, but not sure if semantics will be the same, i.e. will there acutally be N sequental `.isFalse()`'es.

`std::count_if` would not be right here, but your initial `std::distance suggestion` would work. The behavior of negation is a little subtle, which I why I thought the code would be easier to understand as a 2 line for-loop, instead of something that might send someone looking up corner cases in std algorithm man pages.

`span.negated()` is mostly just called as a boolean check to see if were are any negated values in the span and return as quickly as possible. It returns a `size_t` instead of a `bool` because in some cases, the number of negated values is important. A `false` bool value is negated but also negates (skips, clears) any values that came before it, so they won't be included in span iteration, or returned by `GetArgs` or `GetSettingsList`.

Talkless commented on src/util/system.cpp:
```diff
@@ -63,6 +63,7 @@
 #endif
 
 #include <thread>
+#include <univalue.h>
```
Is this the right place to put this include? Shouldn't it be way more up, together with other bitcoin stuff?

Talkless replied:
>  Univalue is technically an external include

Oh, right then.

Talkless commented on src/util/system.cpp:
```diff
@@ -399,49 +312,43 @@ bool ArgsManager::ParseParameters(int argc, const char* const argv[], std::strin
-            return false;
+    bool success = true;
+    if (auto* includes = util::FindKey(m_settings.command_line_options, "includeconf")) {
+        for (const auto& include : util::SettingsSpan(*includes)) {
```
Is it possible for `if` to be true, but loop be empty, i.e. not to loop not a single time (empty span), so `success = false` would be never executed? Or this would be expected?

If it is impossible that `for()` never loops (it always loops at least once), then `bool success` is not needed, `return false` can be returned *after* loop body, and `return true` in the end of the function.

Talkless commented on src/util/settings.h:
```diff
@@ -0,0 +1,82 @@
+    explicit SettingsSpan(const SettingsValue& value) noexcept : SettingsSpan(&value, 1) {}
+    explicit SettingsSpan(const SettingsValue* data, size_t size) noexcept : data(data), size(size) {}
+    explicit SettingsSpan(const std::vector<SettingsValue>& vec) noexcept;
+    const SettingsValue* begin() const; //<! Pointer to first non-negated value.
```
I understand `begin()`, `end()` and `empty` (for std::empty in the future) being lowercase, but shouldn't it be `LastNegated()` and `Negated()` as Bitcoin's PascalCase style requires?

ryanofsky reviewed (commented):
For a guy called @Talkless, you do make a lot of comments... :circus_tent:

Eyerolling puns aside though, if you are still reviewing, I might encourage taking a broader view and looking for impacts of changes here that may be more concerning than style or performance tweaks.

ryanofsky commented:
Updated fdae2210f458d2c4a74fe0c079501bed39ec6404 -> e18bf8dd30259f5c5ca63fdacadebd8b28acb2cc ([pr/mergeset.19](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.19) -> [pr/mergeset.20](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.20), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mergeset.19..pr/mergeset.20)) with test simplifications suggested by @jnewbery removing more `m_override_args` / `m_config_args` usages

jnewbery reviewed (commented):
Still chewing my way through this. I have a few small nits so far. I'll try to finish review early next week.

jnewbery commented on src/util/settings.cpp:
```diff
@@ -0,0 +1,161 @@
+            span = SettingsSpan(*values);
+        }
+    }
+    fn(Source(span).SetConfigFile(/* top_level= */ true));
```
It wasn't immediately obvious to me why `fn()` was being called on `span` even if hasn't been set. Moving this line into the if block causes the tests to fail. Can you add a comment to explain why you're calling this even if no settings were found in the top_level of the config file?

Dianne120 commented:
ok

ryanofsky reviewed (commented):
Updated e18bf8dd30259f5c5ca63fdacadebd8b28acb2cc -> 25c4ab5e0afd02e7dd19ea143f2e087def5f7b87 ([pr/mergeset.20](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.20) -> [pr/mergeset.21](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.21), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mergeset.20..pr/mergeset.21)) with suggested changes from John.

ryanofsky commented on src/test/util_tests.cpp:
```diff
@@ -156,19 +157,58 @@ BOOST_AUTO_TEST_CASE(util_FormatISO8601Date)
     TestArgsManager() { m_network_only_args.clear(); }
     std::map<std::string, std::vector<std::string> >& GetOverrideArgs() { return m_override_args; }
     std::map<std::string, std::vector<std::string> >& GetConfigArgs() { return m_config_args; }
+    bool ParseParameters(int argc, const char* const argv[], std::string& error) {
```
re: https://github.com/bitcoin/bitcoin/pull/15934#discussion_r339186323

> These functions (`TestArgsManager::ParseParameters()`, `TestArgsManager::ReadConfigString()` and `TestArgsManager::UpdateSettings()`) are pretty confusing. 

It was actually easy to update the tests and drop these so I got rid of them in the [pr/mergeset.19](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.19) -> [pr/mergeset.20](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.20) push ([compare](https://github.com/ryanofsky/bitcoin/compare/pr/mergeset.19..pr/mergeset.20))

Talkless commented:
> I might encourage taking a broader view and looking for impacts of changes here that may be more concerning than style or performance tweaks.

Sorry, I can't do more in-depth review. And that's enough for nit-picking, not to be too annoying...

Talkless reviewed (approved):
Concept ACK. Not tested, reviewed only cosmetics.

ryanofsky commented:
> Sorry, I can't do more in-depth review. And that's enough for nit-picking, not to be too annoying...

Thanks for review @Talkless. Your suggestions improved the PR, and having more eyes on code is always reassuring. Aside from suggesting cosmetic improvements, another thing you can try as a reviewer without specialized knowledge of the code is just asking questions. E.g. spend 20 minutes looking at a change, find the thing that seems most confusing or surprising in the code and ask about it. Chances are other people will wonder about the same thing, and usually whatever it is could stand to be clarified or at least explained better, so you learn something and help make the project more accessible, too.

ryanofsky commented:
Updated 25c4ab5e0afd02e7dd19ea143f2e087def5f7b87 -> 69d44f3cc75a68d404ca0e1ca2b4831fd2bac4bb ([pr/mergeset.21](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.21) -> [pr/mergeset.22](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.22), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mergeset.21..pr/mergeset.22)) just using "default section" more and not sometimes calling it "top level"

jnewbery commented:
I've tried really hard to review this over the last few days, but the amount of change in the _Deduplicate settings merge code_ commit has made this very difficult for me. Perhaps if that commit was broken up into smaller pieces that moved the different `ArgsManager` functions to use the settings logic one-by-one, it'd be easier for me to satisfy myself that this isn't changing behaviour.

I also find the new logic in settings.cpp very difficult to follow, where a callback is being passed from the `GetSetting()`/`GetSettingsList()`/`OnlyHasDefaultSectionSetting()` functions into `MergeSettings()`, and that callback is using variables from various places to update `result`.

The new regression test is very nice and helps reassure me that this hasn't changed behaviour, but I unfortunately can't give an code-review ACK.

ryanofsky commented:
I think this PR could use a re-ack or two from previous reviewers, but otherwise should be close to ready for merge, given all the review it's had so far, and the exhaustive tests [`util_ArgsMerge`](https://github.com/bitcoin/bitcoin/blob/deb2327b435925c6a39ca654a79283b8eb6aeb86/src/test/util_tests.cpp#L626-L822) and [`util_ChainMerge`](https://github.com/bitcoin/bitcoin/blob/deb2327b435925c6a39ca654a79283b8eb6aeb86/src/test/util_tests.cpp#L843-L924) added in supporting PRs #15869 and #15988 to guarantee this PR doesn't change behavior.

Here is a list of previous reviews:

promag concept ack 2019-05-02 https://github.com/bitcoin/bitcoin/pull/15934#pullrequestreview-232997353
fanquake concept ack 2019-05-24 https://github.com/bitcoin/bitcoin/pull/15934#issuecomment-495641177
jamesob tested ack 2019-05-28 https://github.com/bitcoin/bitcoin/pull/15934#pullrequestreview-242727281
jamesob reack 2019-05-29 https://github.com/bitcoin/bitcoin/pull/15934#issuecomment-497054493
jamesob reack 2019-06-27 https://github.com/bitcoin/bitcoin/pull/15934#issuecomment-506518739
jamesob reack 2019-06-28 https://github.com/bitcoin/bitcoin/pull/15934#issuecomment-506816948
hebasto concept ack 2019-06-31 https://github.com/bitcoin/bitcoin/pull/15934#issuecomment-516792401
ariard review 2019-10-17 https://github.com/bitcoin/bitcoin/pull/15934#pullrequestreview-302987362
ariard tested ack 2019-10-22 https://github.com/bitcoin/bitcoin/pull/15934#pullrequestreview-305582886
fjahr tested ack 2019-10-23 https://github.com/bitcoin/bitcoin/pull/15934#pullrequestreview-305976230
jonatack tested ack 2019-10-23 https://github.com/bitcoin/bitcoin/pull/15934#pullrequestreview-306047236
fjahr reack 2019-10-23 https://github.com/bitcoin/bitcoin/pull/15934#issuecomment-545675250
talkless review 2019-10-24 https://github.com/bitcoin/bitcoin/pull/15934#pullrequestreview-306757104
jnewbery review 2019-10-25 https://github.com/bitcoin/bitcoin/pull/15934#pullrequestreview-307333248
jnewbery review 2019-10-29 https://github.com/bitcoin/bitcoin/pull/15934#issuecomment-547463678

---

re: https://github.com/bitcoin/bitcoin/pull/15934#issuecomment-547463678 from jnewbery

> I've tried really hard to review this over the last few days, but the amount of change in the _Deduplicate settings merge code_ commit has made this very difficult for me. Perhaps if that commit was broken up into smaller pieces that moved the different `ArgsManager` functions to use the settings logic one-by-one, it'd be easier for me to satisfy myself that this isn't changing behaviour.

I looked into this and it's possible to make the PR do a more gradual transition between the old and the new merge implementations by adding new temporary code and then deleting it within the PR. For a first step (which maybe is sufficient for your purpose of understanding the previous code better) I was able to come up with eadbe7c164454a9faf0992d1605742a5626e729 from [`pr/merge2`](https://github.com/ryanofsky/bitcoin/commits/pr/merge2) which adds an initial commit at something like a halfway point between the new code and old code.

But what you're trying to do strikes me as an unnecessarily painful way to approach the review. The PR is replacing duplicated, sprawled out, buggy, opaque, and completely insane merging code, with a new, slightly less insane, more contained and better documented implementation. There isn't a reason as a reviewer to dig into details of the old implementation when there are regression tests covering all the ArgsManager inputs and outputs to guarantee the new implementation is equivalent. If the new implementation can be improved, I'd like to improve it! If understanding details of how the new and old implementations map on to each other is interesting, I am happy to walk through them. But I don't think a longer, more meandering PR that gets to the same result would be an improvement in this particular case.

> I also find the new logic in settings.cpp very difficult to follow, where a callback is being passed from the `GetSetting()`/`GetSettingsList()`/`OnlyHasDefaultSectionSetting()` functions into `MergeSettings()`, and that callback is using variables from various places to update `result`.

I wonder if would help to rename `MergeSettings` to `ForEachSettingsSource`. This is just using an idiom for looping which is pretty common in other codebases, with `MergeSettings` here looping over fields of a struct. If you were trying to initially understand `for(int i = 1; i < 10; ++i) { code; }` and 
 `ForInterval(1, 10, [&](int i) { code; });` you could easily get hung up on either one, but once you've seen the idiom they should both become pretty straightforward.

> The new regression test is very nice and helps reassure me that this hasn't changed behaviour, but I unfortunately can't give an code-review ACK.

That's ok and thanks for the improvements you suggested before. If there's anything I can do to help, like showing where things correspond in the old and new implementations, or improving the new implementation, I'm happy to make the effort.

jamesob reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/15934/commits/202e19830854550ee4eade53fe4fd3ae323ddbdd

ReACKing this after rereviewing the code commit-by-commit. Admittedly my reading of the code this time around was less comprehensive than my last review -- paging in an understanding of the existing settings code is very difficult and time-consuming -- but I read enough to verify there's nothing malicious here. Given how well covered settings parsing is by tests and the fact that this region of the code is low-risk, I feel good about merging this change.

I built the code locally and repeated the manual tests outlined in https://github.com/bitcoin/bitcoin/pull/15934#pullrequestreview-242727281.

Edit: forgot to mention that the second commit is most easily reviewed in a split view IMO.

jnewbery commented on src/util/system.cpp:
```diff
@@ -899,25 +782,31 @@ bool ArgsManager::ReadConfigFiles(std::string& error, bool ignore_invalid_keys)
         bool emptyIncludeConf;
```
I can't comment on the exact line above (L897), but that comment refers to 'override args' and so should be changed in this PR.

I think the comment and bool name are pretty confusing, and could be made much clearer. Suggestion:

```
        // `-includeconf` cannot be included in the command line arguments except
        // as `-noincludeconf` (which indicates that no conf file should be used).
        bool use_conf_file{true};
        {
            LOCK(cs_args);
            const std::vector<util::SettingsValue>* const includes = util::FindKey(m_settings.command_line_options, "includeconf");
            if (includes) {
                // ParseParameters() fails if a non-negated -includeconf is passed on the command-line
                assert(util::SettingsSpan(*includes).last_negated());
                use_conf_file = false;
            }
        }
        if (use_conf_file) {
```

jnewbery commented on src/util/system.cpp:
```diff
@@ -899,25 +782,31 @@ bool ArgsManager::ReadConfigFiles(std::string& error, bool ignore_invalid_keys)
-                std::vector<std::string> includeconf_net(GetArgs(std::string("-") + chain_id + ".includeconf"));
-                includeconf.insert(includeconf.end(), includeconf_net.begin(), includeconf_net.end());
-            }
+            std::vector<std::string> includeconf;
```
These variable names (which are unchanged by your PR) aren't very helpful. I'd suggest the following renames for a later PR:

`includeconf` -> `conf_file_names`
`to_include` -> `conf_file_name`
`include_conf` -> `conf_file_stream`

jnewbery commented on src/util/system.cpp:
```diff
@@ -933,15 +822,14 @@ bool ArgsManager::ReadConfigFiles(std::string& error, bool ignore_invalid_keys)
+            includeconf.clear();
+            add_includes(chain_id, /* skip= */ chain_includes);
+            add_includes({}, /* skip= */ default_includes);
             {
```
(unchanged but touched by this PR): I don't understand why this needs to be its own scope. Perhaps it can be flattened in a follow-up PR?

jnewbery commented on src/util/settings.cpp:
```diff
@@ -0,0 +1,163 @@
+namespace util {
+namespace {
+
+struct Source {
```
For me, this struct doesn't make things any clearer. An alternative is to pass an enum into the callback function to indicate where the SettingsSpan came from. Take a look at the change here, which also adds a few more comments: https://github.com/jnewbery/bitcoin/commit/fc3e1d4a2b8201cc694aee01a84b2b5e937b99cb

jnewbery reviewed (commented):
utACK 202e19830

I have a couple of suggestions for clarifying the code in some places. Take a look at https://github.com/jnewbery/bitcoin/commits/pr15934.2 and let me know what you think.

ariard commented on src/util/settings.cpp:
```diff
@@ -0,0 +1,163 @@
+    SettingsValue result;
+    MergeSettings(settings, section, name, [&](Source source) {
+        // Weird behavior preserved for backwards compatibility: Apply negated
+        // setting in otherwise ignored sections. A negated value in the
```
So the behavior you're describing is in case of negated value in default section, this result is going to take precedence on value in network section even it's a non-negated one ? If yes saying that network specific options are overriden would be more correct than ignored.

ariard replied:
Thanks, after your explanation, I think comment is good enough (or don't suggestion in mind to make it better)

ariard reviewed (commented):
@ryanofsky I'm curious of more details on some weird behaviors you're documented, it's hard to grasp them or comment may be better. Otherwise my others comments are nits or follow-ups improvements, holding my ACK just on answers to weird behaviors.

Advice to any other reviewer, it's easier to understand PR starting in reverse topological order _Deduplicate settings merge code_ and  then _Add util::Settings struct and helper functions_

ryanofsky reviewed (commented):
Thanks John and Antoine for sticking with this and not giving up on this insanity! I'm adopting every suggestion here and making changes to address all comments.

ryanofsky committed:
Clarify emptyIncludeConf logic

Suggestion from John Newbery <john@johnnewbery.com> in
https://github.com/bitcoin/bitcoin/pull/15934#discussion_r343795528

ryanofsky committed:
Rename includeconf variables for clarity

includeconf -> conf_file_names
to_include -> conf_file_name
include_config -> conf_file_stream

Suggestion from John Newbery <john@johnnewbery.com> in
https://github.com/bitcoin/bitcoin/pull/15934#discussion_r343905138

ryanofsky committed:
Remove includeconf nested scope

Easier to review ignoring whitespace

Suggestion from John Newbery <john@johnnewbery.com> in
https://github.com/bitcoin/bitcoin/pull/15934#discussion_r343806860

ryanofsky commented:
Updated 202e19830854550ee4eade53fe4fd3ae323ddbdd -> 20f74e2fba764299f48bd872c9dc2df86c21e747 ([`pr/mergeset.23`](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.23) -> [`pr/mergeset.24`](https://github.com/ryanofsky/bitcoin/commits/pr/mergeset.24), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mergeset.23..pr/mergeset.24)) with suggestions above.

jnewbery commented on src/util/system.cpp:
```diff
@@ -331,22 +257,9 @@ const std::set<std::string> ArgsManager::GetUnsuitableSectionOnlyArgs() const
-
-        // otherwise, issue a warning
-        unsuitables.insert(arg);
+        if (OnlyHasDefaultSectionSetting(m_settings, m_network, SettingName(arg))) {
```
I agree. The:

```
    for (const auto& arg : gArgs.GetUnsuitableSectionOnlyArgs()) {
        return InitError(strprintf(_("Config setting for %s only applied on %s network when in [%s] section.").translated, arg, network, network));
    }
```

in `AppInitParameterInteraction()` clearly doesn't make sense.

ariard replied:
Yes I'm fine too if we print all errors than return

ariard commented on src/util/system.cpp:
```diff
@@ -961,9 +849,16 @@ bool ArgsManager::ReadConfigFiles(std::string& error, bool ignore_invalid_keys)
+
+    const bool fRegTest = get_net("-regtest");
+    const bool fTestNet = get_net("-testnet");
     const bool is_chain_arg_set = IsArgSet("-chain");
```
> I don't see this. Was it intentional to omit the leading dash before -chain? 

Ah sorry I tested with `chain=0` in config file, but didn't remember that `bitcoind` was pretty liberal with non-sense in config file..

That's said would it change anything to just return `CBaseChainParams::MAIN` at then of this function ? Assuming there is no conflict with other networks, if `-chain` isn't set you're going anyways to start on mainnet.

ryanofsky replied:
> That's said would it change anything to just return `CBaseChainParams::MAIN` at then of this function ? Assuming there is no conflict with other networks, if `-chain` isn't set you're going anyways to start on mainnet.

I might be misunderstanding the suggestion, but if the suggestion is to change `return GetArg("-chain", CBaseChainParams::MAIN);` to `return CBaseChainParams::MAIN;` then that will break usages `bitcoind -chain=test` or `bitcoind -chain=regtest`.

If the question is why the overlapping options exist in the first place, I don't know the answer, but this was done in #16680, and the discussion there probably has some justifications.

ariard replied:
Okay you're right didn't have the `chain=test` syntax in mind.

ariard commented on src/util/settings.cpp:
```diff
@@ -0,0 +1,163 @@
+    MergeSettings(settings, section, name, [&](Source source) {
+        // Weird behavior preserved for backwards compatibility: Apply config
+        // file settings even if negated on command line. Negating a setting on
+        // command line will discard earlier settings on the command line and
```
Thanks, gotcha, I find code easier to grasp moving to `Source::CONFIG_FILE_DEFAULT_SECTION`, etc

ariard commented on src/util/settings.cpp:
```diff
@@ -0,0 +1,163 @@
+        // assigned value instead of last. In general, later settings take
+        // precedence over early settings, but for backwards compatibility in
+        // the config file the precedence is reversed for most settings.
+        const bool reverse_precedence = source.config_file && !get_chain_name;
```
Ah no you're right, nothing to do with `ReadConfigStream` parsing, that's just a weird flag of `GetArgHelper`, without that much rational..

ariard commented on src/util/system.cpp:
```diff
@@ -894,14 +894,19 @@ bool ArgsManager::ReadConfigFiles(std::string& error, bool ignore_invalid_keys)
-        // passed '-noincludeconf' on the command line, in which case we should not include anything
-        bool emptyIncludeConf;
+        // `-includeconf` cannot be included in the command line arguments except
+        // as `-noincludeconf` (which indicates that no conf file should be used).
```
I find comment a bit confusing, `-noincludeconf` means we are only going to discard `includeconf` in config file but not the main config itself (at least I've tested that the current behavior with this patchset)

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/15934#discussion_r344270169

> I find comment a bit confusing, `-noincludeconf` means we are only going to discard `includeconf` in config file but not the main config itself (at least I've tested that the current behavior with this patchset)

`-noincludeconf` just disables the `-includeconf` directive. I wouldn't expect it to disable the whole config file. (I might expect `-noconf` to do that.) So this doesn't seem that confusing to me, but I'm happy to take a suggestion if you have an idea to improve it.

The current text just comes from John's suggestion https://github.com/bitcoin/bitcoin/pull/15934#discussion_r343795528

ariard replied:
nit-picking there, was just to change comment to `(which indicates that no included conf file should be used)`

ariard commented:
ACK 083c954

I've reviewed that's new code is doing what it says to do. I think that weird behaviors are compatibility-maintained but at least if we have a regression they are well-documented and so we'll know where to look, plus as said by James this part of code is low-risk so it's worth moving forward.

Great work Russ, and I'm concept ACK future PRs to remove/squeeze weird behaviors!

jnewbery reviewed (commented):
ACK 083c954b02a4e7d0708349eeaf3bac2b5947fb0e

ryanofsky commented:
Thanks! Will leave PR at 083c954b02a4e7d0708349eeaf3bac2b5947fb0e. Some notes for future followup:

- [ ] https://github.com/bitcoin/bitcoin/pull/15934#discussion_r344291875 - s/conf file/included conf file/
- [ ] https://github.com/bitcoin/bitcoin/pull/15934#discussion_r344271224 - s/, so we clear it here// and early return

jnewbery commented:
Other potential future follow-ups:

- [ ] Don't throw in `GetChainName()`

(https://github.com/bitcoin/bitcoin/blob/8021392b825c74312173f15eb937ba6d4aec3841/src/util/system.cpp#L970). I think the calling code was originally supposed to catch this (eg https://github.com/bitcoin/bitcoin/blob/8021392b825c74312173f15eb937ba6d4aec3841/src/bitcoind.cpp#L88), but this is no longer the case because `GetChainName()` is called in `ReadConfigFiles()`, and so starting with `-regtest -testnet` results in an unhandled exception.

- [ ] Remove `ArgsManagerHelper()`

Here in master: https://github.com/bitcoin/bitcoin/blob/8021392b825c74312173f15eb937ba6d4aec3841/src/util/system.cpp#L165. After this PR there are only two functions in this class. Just move them into `ArgsManager()`. The class was only added to speed up rebuilding (https://github.com/bitcoin/bitcoin/pull/11862#discussion_r173870835), which doesn't seem like a good reason to keep it.

- [ ] Move the `ArgsManager` code into `util/settings` ?

jamesob commented on src/util/settings.cpp:
```diff
@@ -0,0 +1,169 @@
+        // precedence over early settings, but for backwards compatibility in
+        // the config file the precedence is reversed for all settings except
+        // chain name settings.
+        const bool reverse_precedence = (source == Source::CONFIG_FILE_NETWORK_SECTION || source == Source::CONFIG_FILE_DEFAULT_SECTION) && !get_chain_name;
```
These long lines are killin' me but I'll bite my tongue in the interest of a merge.

ryanofsky replied:
Will fix with the other followups, but if it in case its useful https://greasyfork.org/en/scripts/18789-github-toggle-code-wrap seems to work well. (Found through https://github.com/StylishThemes/GitHub-code-wrap)

jamesob reviewed (approved):
ACK 083c954b02a4e7d0708349eeaf3bac2b5947fb0e

Reviewed diff-of-diffs both on Github [and locally](https://gist.github.com/jamesob/778eb138a8fe41148f173c791727dcb8). Only changes since my last ACK are those discussed above. Nice job persevering, @ryanofsky, and thanks to @jnewbery @ariard for good feedback.

ryanofsky commented:
I've been working on some followups for this PR since it's been merged. #17473 has various refactoring cleanups suggested above, and #17508 is tracking cleanup of confusing settings behaviors.

Going through all the comments in this issue, I think the only suggestions left not addressed by #17473 and #17508 are:

- Fixing `return InitError` in a loop: https://github.com/bitcoin/bitcoin/pull/15934#discussion_r344252898
- Simplifying `-includeconf` errors: https://github.com/bitcoin/bitcoin/pull/15934#discussion_r344271224
- Fixing uncaught exceptions thrown by `GetChainName`: https://github.com/bitcoin/bitcoin/pull/15934#issuecomment-551969778
- Moving ArgsManager settings code from system.cpp to settings.cpp: https://github.com/bitcoin/bitcoin/pull/15934#issuecomment-551969778

I don't have plans to work on these last few things

ariard commented:
Maybe tag them good_first_issue or up_to_grabs to let people find them

MarcoFalke commented:
Hidden by GitHub, but the "Up for grabs" refers to https://github.com/bitcoin/bitcoin/pull/15934#issuecomment-555100626
